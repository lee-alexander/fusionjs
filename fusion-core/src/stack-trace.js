/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

export function captureStackTrace(caller: Function): string {
  // For monitoring in production, use a single format independent of browser
  if (__BROWSER__ && !__DEV__) {
    return new Error().stack;
  } else {
    if ('captureStackTrace' in Error) {
      const err = {};
      Error.captureStackTrace(err, caller);
      return err.stack;
    } else {
      return new Error().stack;
    }
  }
}

export class DIError extends Error {
  link: ?string;
  constructor({
    message,
    errorDoc,
    caller,
    stack,
  }: {
    message: string,
    errorDoc?: string,
    caller?: Function,
    stack?: string,
  }) {
    super(message);
    if (errorDoc) {
      this.link = `https://github.com/fusionjs/fusionjs/tree/master/errors/${errorDoc}.md`;
    }
    if (__NODE__ || __DEV__) {
      if (caller && 'captureStackTrace' in Error) {
        Error.captureStackTrace(this, caller);
      }
    }
    if (stack) {
      // Replace the DIError stack trace with the supplied stack
      // Assumes `message` contains no new lines
      const diErrorMessage = this.stack.split('\n')[0];
      const stackOverride = stack.split('\n');
      stackOverride[0] = diErrorMessage;
      this.stack = stackOverride.join('\n');
    }
  }
}
