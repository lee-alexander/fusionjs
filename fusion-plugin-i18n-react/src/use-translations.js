/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import {useContext, useMemo} from 'react';
import {I18nContext} from './plugin.js';

export function useTranslations() {
  const i18n = useContext(I18nContext);

  return useMemo(() => i18n.translate.bind(i18n), [i18n]);
}
