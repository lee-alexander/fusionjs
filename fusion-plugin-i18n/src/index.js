/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import {createToken} from 'fusion-core';
import type {FusionPlugin, Token} from 'fusion-core';

import serverPlugin from './node';
import clientPlugin, {HydrationStateToken} from './browser';
import serverLoader from './loader';
import type {
  I18nDepsType,
  I18nServiceType,
  TranslationsObjectType,
  TranslateFuncType,
} from './types.js';
import {I18nLoaderToken} from './tokens.js';

const I18nToken: Token<I18nServiceType> = createToken('I18nToken');

const plugin: FusionPlugin<I18nDepsType, I18nServiceType> = __NODE__
  ? serverPlugin
  : clientPlugin;

const createI18nLoader: typeof serverLoader = __NODE__
  ? serverLoader
  : (void 0: any);

export type {
  I18nDepsType,
  I18nServiceType,
  TranslationsObjectType,
  TranslateFuncType,
};
export default plugin;
export {I18nToken, I18nLoaderToken, HydrationStateToken, createI18nLoader};
