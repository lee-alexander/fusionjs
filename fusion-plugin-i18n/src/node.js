/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/* eslint-env node */

import {Locale} from 'locale';

import {createPlugin, memoize, html} from 'fusion-core';
import type {FusionPlugin, Context} from 'fusion-core';
import {UniversalEventsToken} from 'fusion-plugin-universal-events';
import bodyparser from 'koa-bodyparser';
import querystring from 'querystring';

import {I18nLoaderToken} from './tokens.js';
import createLoader from './loader.js';
import type {
  I18nDepsType,
  I18nServiceType,
  TranslationsObjectType,
  IEmitter,
} from './types.js';

// exported for testing
export function matchesLiteralSections(literalSections: Array<string>) {
  return (translation: string) => {
    let lastMatchIndex = 0;

    if (literalSections.length === 1) {
      const literal = literalSections[0];
      return literal !== '' && translation === literal;
    }

    return literalSections.every((literal, literalIndex) => {
      if (literal === '') {
        // literal section either:
        // - starts/ends the literal
        // - is the result of two adjacent interpolations
        return true;
      } else if (literalIndex === 0 && translation.startsWith(literal)) {
        lastMatchIndex += literal.length;
        return true;
      } else if (
        literalIndex === literalSections.length - 1 &&
        translation.endsWith(literal)
      ) {
        return true;
      } else {
        // start search from `lastMatchIndex`
        const matchIndex = translation.indexOf(literal, lastMatchIndex);
        if (matchIndex !== -1) {
          lastMatchIndex = matchIndex + literal.length;
          return true;
        }
      }
      // matching failed
      return false;
    });
  };
}

function getKeysFromContext(ctx: Context): string[] {
  if (ctx.request.body && Array.isArray(ctx.request.body)) {
    return (ctx.request.body: any);
  }

  const querystringParams = querystring.parse(ctx.querystring);
  if (querystringParams.keys) {
    try {
      const keys = JSON.parse(querystringParams.keys);
      return Array.isArray(keys) ? keys : [];
    } catch (e) {
      return [];
    }
  }

  return [];
}

type PluginType = FusionPlugin<I18nDepsType, I18nServiceType>;
const pluginFactory: () => PluginType = () =>
  createPlugin({
    deps: {
      loader: I18nLoaderToken.optional,
      events: UniversalEventsToken.optional,
    },
    provides: ({loader, events}) => {
      class I18n {
        translations: TranslationsObjectType;
        locale: string | Locale;
        emitter: ?IEmitter;

        constructor(ctx) {
          if (!loader) {
            loader = createLoader();
          }
          const {translations, locale} = loader.from(ctx);
          this.emitter = events && events.from(ctx);
          this.translations = translations;
          this.locale = locale;
        }
        async load() {} //mirror client API
        translate(key, interpolations = {}) {
          const template = this.translations[key];

          if (typeof template !== 'string') {
            this.emitter && this.emitter.emit('i18n-translate-miss', {key});
            return key;
          }

          return template.replace(/\${(.*?)}/g, (_, k) =>
            interpolations[k] === void 0
              ? '${' + k + '}'
              : String(interpolations[k])
          );
        }
      }

      const service = {from: memoize(ctx => new I18n(ctx))};
      return service;
    },
    middleware: (_, plugin) => {
      // TODO(#4) refactor: this currently depends on babel plugins in framework's webpack config.
      // Ideally these babel plugins should be part of this package, not hard-coded in framework core
      const chunkTranslationMap = require('../chunk-translation-map');
      const parseBody = bodyparser();

      return async (ctx, next) => {
        if (ctx.element) {
          await next();
          const i18n = plugin.from(ctx);

          // get the webpack chunks that are used and serialize their translations
          const chunks: Array<string | number> = [
            ...ctx.syncChunks,
            ...ctx.preloadChunks,
          ];
          const translations = {};
          const possibleTranslations = i18n.translations
            ? Object.keys(i18n.translations)
            : [];
          chunks.forEach(id => {
            const keys = Array.from(
              chunkTranslationMap.translationsForChunk(id)
            );
            keys.forEach(key => {
              if (Array.isArray(key)) {
                const matches = possibleTranslations.filter(
                  matchesLiteralSections(key)
                );
                for (const match of matches) {
                  translations[match] =
                    i18n.translations && i18n.translations[match];
                }
              } else {
                translations[key] = i18n.translations && i18n.translations[key];
              }
            });
          });
          // i18n.locale is actually a locale.Locale instance
          if (!i18n.locale) {
            throw new Error('i18n.locale was empty');
          }
          const localeCode =
            typeof i18n.locale === 'string' ? i18n.locale : i18n.locale.code;
          const serialized = JSON.stringify({localeCode, translations});
          const script = html`
            <script type="application/json" id="__TRANSLATIONS__">
              ${serialized}
            </script>
          `; // consumed by ./browser
          ctx.template.body.push(script);

          // set HTML lang tag as a hint for signal screen readers to switch to the
          // recommended language.
          ctx.template.htmlAttrs.lang = localeCode;
        } else if (ctx.path === '/_translations') {
          const i18n = plugin.from(ctx);
          try {
            await parseBody(ctx, () => Promise.resolve());
          } catch (e) {
            ctx.request.body = [];
          }
          const keys = getKeysFromContext(ctx);
          const possibleTranslations = i18n.translations
            ? Object.keys(i18n.translations)
            : [];
          const translations = keys.reduce((acc, key) => {
            if (Array.isArray(key)) {
              const matches = possibleTranslations.filter(
                matchesLiteralSections(key)
              );
              for (const match of matches) {
                acc[match] = i18n.translations && i18n.translations[match];
              }
            } else {
              acc[key] = i18n.translations && i18n.translations[key];
            }
            return acc;
          }, {});
          ctx.body = translations;
          ctx.set('cache-control', 'public, max-age=3600'); // cache translations for up to 1 hour
          return next();
        } else {
          return next();
        }
      };
    },
  });

export default ((__NODE__ && pluginFactory(): any): PluginType);
