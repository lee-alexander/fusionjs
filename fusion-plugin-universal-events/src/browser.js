/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/* eslint-env browser */
import {createPlugin} from 'fusion-core';
import type {FusionPlugin} from 'fusion-core';
import {FetchToken} from 'fusion-tokens';
import type {Fetch} from 'fusion-tokens';
import Emitter from './emitter.js';
import type {
  IEmitter,
  UniversalEventsPluginDepsType as DepsType,
  BatchStorage,
} from './types.js';
import {
  UniversalEventsBatchStorageToken,
  localBatchStorage,
} from './storage/index.js';

export class UniversalEmitter extends Emitter {
  flush: any;
  fetch: Fetch;
  interval: any;
  storage: BatchStorage;
  limit: number;
  isFlushInProgress = false;
  hasFlushBeenScheduled = false;

  constructor(
    fetch: Fetch,
    storage: BatchStorage,
    interval?: number = 5000,
    limit?: number = 1000
  ): void {
    super();
    //privates
    this.storage = storage;
    this.flush = this.flushInternal.bind(this);
    this.fetch = fetch;
    this.setFrequency(interval);
    this.limit = limit;
    window.addEventListener('visibilitychange', this.flushBeforeTerminated);
  }
  setFrequency(frequency: number): void {
    window.clearInterval(this.interval);
    this.interval = setInterval(this.flush, frequency);
  }
  emit(type: mixed, payload: mixed): void {
    payload = super.mapEvent(type, payload);
    super.handleEvent(type, payload);
    this.storage.add({type, payload});
  }
  // match server api
  from(): UniversalEmitter {
    return this;
  }
  flushBeforeTerminated = () =>
    document.visibilityState === 'hidden' && this.flushInternal();
  async flushInternal(): Promise<void> {
    if (!this.startFlush()) {
      return;
    }
    // this is workaround to 'get' items and keep them in the storage
    const items = this.storage.getAndClear(this.limit);
    this.storage.addToStart(...items);

    if (items.length === 0) {
      this.finishFlush();
      return;
    }
    try {
      const res = await this.fetch('/_events', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({items}),
      });

      if (res.ok) {
        // clear only events that were sent, preserve ones that might be added while request was executed
        this.storage.getAndClear(items.length);
        this.finishFlush();
      } else {
        // If the server responds with a 413, it means the size of the payload was too large.
        // We handle this by cutting our limit in half for the next attempt.
        if (res.status === 413) {
          this.limit = this.limit / 2;
        }
        this.finishFlush();
      }
    } catch (e) {
      // do nothing here, items are still in the storage and will be sent on the next attempt
      this.finishFlush();
    }
  }
  teardown(): void {
    window.removeEventListener('visibilitychange', this.flushBeforeTerminated);
    clearInterval(this.interval);
    this.interval = null;
  }
  startFlush() {
    if (this.isFlushInProgress) {
      this.hasFlushBeenScheduled = true;
      return false;
    }
    this.isFlushInProgress = true;
    return true;
  }
  finishFlush() {
    this.isFlushInProgress = false;
    if (this.hasFlushBeenScheduled) {
      this.hasFlushBeenScheduled = false;
      this.flushInternal();
    }
  }
}

const plugin =
  __BROWSER__ &&
  createPlugin({
    deps: {
      fetch: FetchToken,
      storage: UniversalEventsBatchStorageToken.optional,
    },
    provides: ({fetch, storage}) => {
      return new UniversalEmitter(fetch, storage || localBatchStorage);
    },
    cleanup: async emitter => {
      return emitter.teardown();
    },
  });

export default ((plugin: any): FusionPlugin<DepsType, IEmitter>);
