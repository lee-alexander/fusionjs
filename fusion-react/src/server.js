/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/* eslint-env node */
import * as React from 'react';
import {renderToString} from 'react-dom/server';
import type {Logger} from 'fusion-tokens';

export default (el: React.Element<*>, logger?: Logger) => {
  try {
    return `<div id='root'>${renderToString(el)}</div>`;
  } catch (e) {
    if (__DEV__) {
      console.error(
        'Server-side render failed. Falling back to client-side render'
      );
    }
    logger && logger.error('SSR Failed with Error', e);
    return '<div id="root" data-fusion-render="client"></div>';
  }
};
