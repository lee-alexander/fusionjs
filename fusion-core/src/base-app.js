/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
import {createPlugin} from './create-plugin';
import {createToken, TokenType, TokenImpl} from './create-token';
import {
  ElementToken,
  RenderToken,
  SSRDeciderToken,
  RouteTagsToken,
} from './tokens';
import {SSRDecider} from './plugins/ssr';
import RouteTagsPlugin from './plugins/route-tags';
import {captureStackTrace, DIError} from './stack-trace.js';

import type {aliaser, cleanupFn, FusionPlugin, Token} from './types.js';

interface Register<T> {
  (Token<T>, mixed): aliaser<Token<T>>;
  (mixed): aliaser<Token<T>>;
}

class FusionApp {
  constructor(el: Element | string, render: any => any) {
    this.registered = new Map(); // getTokenRef(token) -> {value, aliases, enhancers}
    this.enhancerToToken = new Map(); // enhancer -> token
    this._dependedOn = new Set();
    this.plugins = []; // Token
    this.cleanups = [];
    el && this.register(ElementToken, el);
    render && this.register(RenderToken, render);
    this.register(SSRDeciderToken, SSRDecider);
    this.register(RouteTagsToken, RouteTagsPlugin);
  }

  // eslint-disable-next-line
  registered: Map<
    any,
    {
      aliases?: Map<any, any>,
      enhancers?: Array<any>,
      token: any,
      value?: FusionPlugin<*, *>,
    }
  >;
  enhancerToToken: Map<any, any>;
  plugins: Array<any>;
  cleanups: Array<cleanupFn>;
  renderer: any;
  _getService: any => any;
  _dependedOn: Set<any>;

  register: Register<*> = <T>(tokenOrValue, maybeValue) => {
    const hasToken = tokenOrValue instanceof TokenImpl;
    const token = hasToken
      ? ((tokenOrValue: any): Token<T>)
      : createToken('UnnamedPlugin');
    const value: any = hasToken ? maybeValue : tokenOrValue;
    if (!hasToken && (value == null || !value.__plugin__)) {
      throw new DIError({
        message: __DEV__
          ? `Cannot register ${String(
              tokenOrValue
            )} without a token. Did you accidentally register a ${
              __NODE__ ? 'browser' : 'server'
            } plugin on the ${__NODE__ ? 'server' : 'browser'}?`
          : 'Invalid configuration registration',
        errorDoc: 'value-without-token',
        caller: this.register,
      });
    }
    // the renderer is a special case, since it needs to be always run last
    if (token === RenderToken) {
      this.renderer = value;
      const alias = () => {
        throw new DIError({
          message: 'Aliasing for RenderToken not supported',
          caller: alias,
        });
      };
      return {alias};
    }
    token.stacks.push({
      type: 'register',
      stack: captureStackTrace(this.register),
    });
    if (value && value.__plugin__) {
      token.stacks.push({type: 'plugin', stack: value.stack});
    }
    return this._register(token, value);
  };
  _register<TResolved>(token: Token<TResolved>, value: *) {
    this.plugins.push(token);
    const {aliases, enhancers} = this.registered.get(getTokenRef(token)) || {
      aliases: new Map(),
      enhancers: [],
    };
    if (value && value.__plugin__) {
      if (value.deps) {
        Object.values(value.deps).forEach(token =>
          this._dependedOn.add(getTokenRef(token))
        );
      }
    }
    this.registered.set(getTokenRef(token), {
      value,
      aliases,
      enhancers,
      token,
    });
    const alias = (sourceToken, destToken) => {
      const stack = captureStackTrace(alias);
      sourceToken.stacks.push({type: 'alias-from', stack});
      destToken.stacks.push({type: 'alias-to', stack});
      this._dependedOn.add(getTokenRef(destToken));
      if (aliases) {
        aliases.set(getTokenRef(sourceToken), destToken);
      }
      return {alias};
    };
    return {alias};
  }
  middleware(deps: *, middleware: *) {
    if (middleware === undefined) {
      middleware = () => deps;
    }
    this.register(createPlugin({deps, middleware}));
  }
  enhance<TResolved>(token: Token<TResolved>, enhancer: Function) {
    token.stacks.push({
      type: 'enhance',
      stack: captureStackTrace(this.enhance),
    });
    const {value, aliases, enhancers} = this.registered.get(
      getTokenRef(token)
    ) || {
      aliases: new Map(),
      enhancers: [],
      value: undefined,
    };
    this.enhancerToToken.set(enhancer, token);

    if (enhancers && Array.isArray(enhancers)) {
      enhancers.push(enhancer);
    }
    this.registered.set(getTokenRef(token), {
      value,
      aliases,
      enhancers,
      token,
    });
  }
  cleanup() {
    return Promise.all(this.cleanups.map(fn => fn()));
  }
  resolve<TResolved>() {
    if (!this.renderer) {
      throw new Error('Missing registration for RenderToken');
    }
    this._register(RenderToken, this.renderer);
    const resolved = new Map(); // Token.ref || Token => Service
    const nonPluginTokens = new Set(); // Token
    const resolving = new Set(); // Token.ref || Token
    const registered = this.registered; // Token.ref || Token -> {value, aliases, enhancers}
    const resolvedPlugins = []; // Plugins
    const appliedEnhancers = [];
    const resolveToken = (token: Token<TResolved>, tokenAliases) => {
      // Base: if we have already resolved the type, return it
      if (tokenAliases && tokenAliases.has(getTokenRef(token))) {
        const newToken = tokenAliases.get(getTokenRef(token));
        if (newToken) {
          token = newToken;
        }
      }
      if (resolved.has(getTokenRef(token))) {
        return resolved.get(getTokenRef(token));
      }

      // Base: if currently resolving the same type, we have a circular dependency
      if (resolving.has(getTokenRef(token))) {
        const registerStack = token.stacks.find(t => t.type === 'register');
        throw new DIError({
          message: `Cannot resolve circular dependency: ${token.name}`,
          errorDoc: 'circular-dependencies',
          stack: registerStack && registerStack.stack,
        });
      }

      // Base: the type was never registered, throw error or provide undefined if optional
      let {value, aliases, enhancers} =
        registered.get(getTokenRef(token)) || {};
      if (value === undefined) {
        // Early return if token is optional
        const isOptional =
          token instanceof TokenImpl && token.type === TokenType.Optional;
        if (isOptional && (!enhancers || !enhancers.length)) {
          return;
        }
        const dependents = Array.from(this.registered.entries());

        /**
         * Iterate over the entire list of dependencies and find all
         * dependencies of a given token.
         */
        const findDependentTokens = () => {
          return dependents
            .filter(entry => {
              if (!entry[1].value || !entry[1].value.deps) {
                return false;
              }
              return Object.values(entry[1].value.deps).includes(token);
            })
            .map(entry => entry[1].token.name);
        };
        const findDependentEnhancers = () => {
          return appliedEnhancers
            .filter(([, provides]) => {
              if (!provides || !provides.deps) {
                return false;
              }
              return Object.values(provides.deps).includes(token);
            })
            .map(([enhancer]) => {
              const enhancedToken = this.enhancerToToken.get(enhancer);
              return `EnhancerOf<${
                enhancedToken ? enhancedToken.name : '(unknown)'
              }>`;
            });
        };
        const dependentTokens = [
          ...findDependentTokens(),
          ...findDependentEnhancers(),
        ];

        const duplicates = this.plugins
          .filter(p => p.name === token.name)
          .map(p => {
            const stack = p.stacks.find(t => t.type === 'token');
            return stack.stack;
          });
        const tokenStack = token.stacks.find(t => t.type === 'token');
        if (duplicates.length) {
          // Note: Update when string token equality is implemented
          throw new DIError({
            message: `Missing registration for token "${token.name}". Other tokens with this name have been registered`,
            stack: tokenStack && tokenStack.stack,
            errorDoc: 'duplicate-token-names',
          });
        } else {
          const dependentList = dependentTokens
            .map(token => `"${token}"`)
            .join(', ');
          const plural = dependentTokens.length > 1 ? 's' : '';
          throw new DIError({
            message: `Missing registration for token "${token.name}". Token is required dependency of plugins registered to ${dependentList} token${plural}`,
            stack: tokenStack && tokenStack.stack,
            errorDoc: 'missing-registration',
          });
        }
      }

      // Recursive: get the registered type and resolve it
      resolving.add(getTokenRef(token));

      function resolvePlugin(plugin) {
        const registeredDeps = (plugin && plugin.deps) || {};
        const resolvedDeps = {};
        for (const key in registeredDeps) {
          const registeredToken = registeredDeps[key];
          resolvedDeps[key] = resolveToken(registeredToken, aliases);
        }
        // `provides` should be undefined if the plugin does not have a `provides` function
        let provides =
          plugin && plugin.provides ? plugin.provides(resolvedDeps) : undefined;
        if (plugin && plugin.middleware) {
          resolvedPlugins.push(plugin.middleware(resolvedDeps, provides));
        }
        return provides;
      }

      let provides = value;
      if (value && value.__plugin__) {
        provides = resolvePlugin(provides);
        if (value.cleanup) {
          this.cleanups.push(function() {
            return typeof value.cleanup === 'function'
              ? value.cleanup(provides)
              : Promise.resolve();
          });
        }
      } else {
        nonPluginTokens.add(token);
      }

      if (enhancers && enhancers.length) {
        enhancers.forEach(e => {
          let nextProvides = e(provides);
          appliedEnhancers.push([e, nextProvides]);
          if (nextProvides && nextProvides.__plugin__) {
            // if the token has a plugin enhancer, allow it to be registered with no dependents
            nonPluginTokens.delete(token);
            if (nextProvides.deps) {
              Object.values(nextProvides.deps).forEach(token =>
                this._dependedOn.add(getTokenRef(token))
              );
            }
            nextProvides = resolvePlugin(nextProvides);
          }
          provides = nextProvides;
        });
      }
      resolved.set(getTokenRef(token), provides);
      resolving.delete(getTokenRef(token));
      return provides;
    };

    for (let i = 0; i < this.plugins.length; i++) {
      resolveToken(this.plugins[i]);
    }
    for (const token of nonPluginTokens) {
      if (
        token !== ElementToken &&
        token !== RenderToken &&
        !this._dependedOn.has(getTokenRef(token))
      ) {
        const registerStack = token.stacks.find(t => t.type === 'register');
        throw new DIError({
          message: `Registered token without depending on it: ${token.name}`,
          errorDoc: 'registered-without-depending',
          stack: registerStack && registerStack.stack,
        });
      }
    }

    this.plugins = resolvedPlugins;
    this._getService = token => resolved.get(getTokenRef(token));
  }
  getService<TResolved>(token: Token<TResolved>): any {
    if (!this._getService) {
      throw new DIError({
        message: 'Cannot get service from unresolved app',
        caller: this.getService,
      });
    }
    return this._getService(token);
  }
}

/* Helper functions */
function getTokenRef(token) {
  if (token instanceof TokenImpl) {
    return token.ref;
  }
  return token;
}

export default FusionApp;
