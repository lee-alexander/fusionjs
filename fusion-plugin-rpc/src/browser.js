/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/* eslint-env browser */

import {createPlugin, memoize, type Context} from 'fusion-core';
import {UniversalEventsToken} from 'fusion-plugin-universal-events';
import {I18nToken} from 'fusion-plugin-i18n';
import {FetchToken} from 'fusion-tokens';
import type {Fetch} from 'fusion-tokens';

import {type HandlerType, RPCHandlersConfigToken} from './tokens.js';
import type {RPCPluginType, IEmitter, RPCConfigType} from './types.js';
import {formatApiPath} from './utils.js';

type InitializationOpts = {
  fetch: Fetch,
  emitter: IEmitter,
  localeCode: string,
  rpcConfig: ?RPCConfigType,
};

const statKey = 'rpc:method-client';

class RPC {
  ctx: ?Context;
  emitter: ?IEmitter;
  handlers: ?HandlerType;
  localeCode: ?string;
  fetch: ?Fetch;
  config: ?RPCConfigType;
  apiPath: string;
  constructor({fetch, emitter, rpcConfig, localeCode}: InitializationOpts) {
    this.fetch = fetch;
    this.config = rpcConfig || {};
    this.emitter = emitter;
    this.localeCode = localeCode;

    this.apiPath = formatApiPath(
      rpcConfig && rpcConfig.apiPath ? rpcConfig.apiPath : 'api'
    );
  }

  request<TArgs, TResult>(
    rpcId: string,
    args: TArgs,
    headers: ?{[string]: string},
    options: ?RequestOptions
  ): Promise<TResult> {
    if (!this.fetch) {
      throw new Error('fusion-plugin-rpc requires `fetch`');
    }
    if (!this.emitter) {
      throw new Error('Missing emitter registered to UniversalEventsToken');
    }
    const fetch = this.fetch;
    const emitter = this.emitter;
    const apiPath = this.apiPath;

    const startTime = Date.now();
    const localeParam = this.localeCode ? `?localeCode=${this.localeCode}` : '';

    return fetch(
      `${apiPath}${rpcId}${localeParam}`,
      args instanceof FormData
        ? {
            ...options,
            method: 'POST',
            headers: {
              // Content-Type will be set automatically
              ...(headers || {}),
            },
            body: args,
          }
        : {
            ...options,
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
              ...(headers || {}),
            },
            body: JSON.stringify(args || {}),
          }
    )
      .then(r => r.json())
      .then(args => {
        const {status, data} = args;
        if (status === 'success') {
          emitter.emit(statKey, {
            method: rpcId,
            status: 'success',
            timing: Date.now() - startTime,
          });
          return data;
        } else {
          emitter.emit(statKey, {
            method: rpcId,
            error: data,
            status: 'failure',
            timing: Date.now() - startTime,
          });
          return Promise.reject(data);
        }
      });
  }
}

const pluginFactory: () => RPCPluginType = () =>
  createPlugin({
    deps: {
      fetch: FetchToken,
      emitter: UniversalEventsToken,
      i18n: I18nToken.optional,
      rpcConfig: RPCHandlersConfigToken.optional,
    },
    provides: deps => {
      const {fetch = window.fetch, emitter, rpcConfig, i18n} = deps;

      return {
        from: memoize(ctx => {
          const locale = (i18n && i18n.from(ctx).locale) || '';
          const localeCode = typeof locale === 'string' ? locale : locale.code;
          return new RPC({
            fetch,
            emitter,
            rpcConfig,
            localeCode,
          });
        }),
      };
    },
  });

export default ((__BROWSER__ && pluginFactory(): any): RPCPluginType);
