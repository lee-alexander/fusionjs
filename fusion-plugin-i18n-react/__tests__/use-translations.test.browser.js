// @flow

import React from 'react';
import {mount} from 'enzyme';

import {useTranslations} from '../src/index';
import {I18nContext} from '../src/plugin.js';

test('useTranslations() hook', () => {
  function Foo() {
    const translate = useTranslations();
    return <div>{translate('foo')}</div>;
  }

  const mockI18n = {
    async load() {},
    localeCode: 'fr_CA',
    translate(str) {
      return `${str} bar baz`;
    },
  };

  expect(
    mount(
      <I18nContext.Provider value={mockI18n}>
        <Foo />
      </I18nContext.Provider>
    ).html()
  ).toMatchSnapshot();
});
