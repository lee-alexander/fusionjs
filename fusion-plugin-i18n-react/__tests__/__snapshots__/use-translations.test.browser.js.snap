// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTranslations() hook 1`] = `"<div>foo bar baz</div>"`;
